sklearn/feature_selection/_sequential.py

```diff
@@ -175,18 +181,34 @@ def fit(self, X, y):
             self.n_features_to_select_ if self.direction == 'forward'
             else n_features - self.n_features_to_select_
         )
-        for _ in range(n_iterations):
-            new_feature_idx = self._get_best_new_feature(cloned_estimator, X,
-                                                         y, current_mask)
-            current_mask[new_feature_idx] = True
 
-        if self.direction == 'backward':
-            current_mask = ~current_mask
+        if self.tol is None:
+            for _ in range(n_iterations):
+                new_feature_idx, new_score = self._get_best_new_feature_score(cloned_estimator, X,
+                                                             y, current_mask)
+                current_mask[new_feature_idx] = True
+
+            if self.direction == 'backward':
+                current_mask = ~current_mask
+        else:
+            old_score = 0
+            for _ in range(n_iterations):
+                new_feature_idx, new_score = self._get_best_new_feature_score(cloned_estimator, X,
+                                                             y, current_mask)
+                if new_score - self.tol < old_score:
```

Comment:
If should be possible to move the `if self.tol is None:` condition inside the loop, here to avoid code duplication:

```suggestion
                if self.tol is not None and new_score - self.tol < old_score:
```

Reply:
fix it on https://github.com/scikit-learn/scikit-learn/pull/20145/commits/1c614a8f4246f405a5868133f4cb9d57fd069d0b